url3472127e1f_SergeyKrivohatskiy_tiny_database_tgz-pJ8-ru_spbau_tinydb_bufferManager_IndexTestJ8/src/ru/spbau/tinydb/table/TableBase.java

```diff
@@ -0,0 +1,272 @@
+        bitSetSize = (recordsPerPage + recordsPerPage + 7) / 8;
+        loadHeader();
+    }
+
+
+    private void writeHeader() {
+        try (BufferView firstPageView = bufferManager.getPage(firstPage)) {
+            firstPageView.setInt(0, firstNotFullPage);
+            firstPageView.setInt(4, firstFullPage);
+            firstPageView.setInt(8, rowsCount);
+            firstPageView.setChanged();
+        }
+
+    }
+
+    private void loadHeader() {
+        try (BufferView firstPageView = bufferManager.getPage(firstPage)) {
+            firstNotFullPage = firstPageView.getInt(0);
+            firstFullPage = firstPageView.getInt(4);
+            rowsCount = firstPageView.getInt(8);
+        }
+    }
+
+    public int insert(byte[] row) throws ExecutionException {
+        if(row.length != recordSize) {
+            throw new IllegalArgumentException();
+        }
+        BufferView pageView;
+        if(firstNotFullPage == 0) {
+            firstNotFullPage = bufferManager.getFreePage();
+            writeHeader();
+            pageView = bufferManager.getPage(firstNotFullPage);
+            pageView.setChanged();
+        } else {
+            pageView = bufferManager.getPage(firstNotFullPage);
+        }
+        int pageIndex = firstNotFullPage;
+
+
+        // writing row
+        BufferView bitSetView = pageView.getSubView(POINTERS_SIZE, bitSetSize);
+        BufferView recordsView = pageView.getSubView(POINTERS_SIZE + bitSetSize, recordsPerPage * recordSize);
+        int recordPosition;
+        // Hand made BitSet
+        for(recordPosition = 0; checkBit(bitSetView, recordPosition); recordPosition ++) {
+            if(recordPosition >= recordsPerPage) {
+                throw new RuntimeException("No empty position when page is not full");
+            }
+        }
+        setBit(bitSetView, recordPosition);
+        recordsView.setBytes(recordPosition * recordSize, row);
+        recordsView.close();
+
+        boolean bitSetFull = true;
+        for(int i = 0; i < recordsPerPage; i ++) {
+            if(!checkBit(bitSetView, i)) {
+                bitSetFull = false;
+                break;
+            }
+        }
+        bitSetView.close();
+        if(bitSetFull) {
+            // move to full
+            int prevNotFull = pageView.getInt(0);
+            int nextNotFull = pageView.getInt(4);
+            if(prevNotFull != 0) {
+                BufferView prevView = bufferManager.getPage(prevNotFull);
+                prevView.setInt(4, nextNotFull);
+                prevView.setChanged();
+                prevView.close();
+            } else {
+                firstNotFullPage = 0;
+                writeHeader();
+            }
+            if(nextNotFull != 0) {
+                BufferView nextView = bufferManager.getPage(nextNotFull);
+                nextView.setInt(0, prevNotFull);
+                nextView.setChanged();
+                nextView.close();
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 111:116, 120:125.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.